File: fsharp/bob/Bob.fs
module Bob

open System.Text.RegularExpressions

let isEmpty (input: string): bool = 
    if input = "" then true
    else not (Regex.IsMatch(input, "\S"))

let isUpper (input: string): bool = 
    Regex.IsMatch(input, "[A-Z]") && not (Regex.IsMatch(input, "[a-z]"))

let isQuestion (input: string): bool = 
    Regex.IsMatch(input, "\?\s*$")

let response (input: string): string = 
    let empty = isEmpty input
    let yell = not empty && isUpper input
    let question = not empty && isQuestion input

    match (yell, question) with
    | (true, true)  -> "Calm down, I know what I'm doing!"
    | (true, false) -> "Whoa, chill out!"
    | (false, true) -> "Sure."
    | (false, false) -> 
            if empty then "Fine. Be that way!"
            else "Whatever."

File: python/bob/bob.py
import re


def response(hey_bob):
    def is_empty(input):
        return not re.search(r"\S", input)
    
    def is_upper(input):
        return re.search(r"[A-Z]", input) and not re.search(r"[a-z]", input) 
    
    def is_question(input):
        return re.search(r"\?\s*$", input)

    if is_empty(hey_bob):
        return "Fine. Be that way!"

    if is_upper(hey_bob):
        if is_question(hey_bob):
            resp = "Calm down, I know what I'm doing!"
        else:
            resp = "Whoa, chill out!"
    else:
        if is_question(hey_bob):
            resp = "Sure."
        else:
            resp = "Whatever."
    return resp
